internal/operator/controller.go

```diff
@@ -0,0 +1,473 @@
+
+	if newc.DeletionTimestamp != nil {
+		_, ok := c.clusters[newc.UID]
+		if !ok {
+			return nil
+		}
+		c.mu.Lock()
+		delete(c.clusters, newc.UID)
+		c.mu.Unlock()
+		log.Debugf("Deleting '%s/%s' cluster (uid=%s)", newc.Namespace, newc.Name, newc.UID)
+		return nil
+	}
+
+	defer func() {
+		c.mu.Lock()
+		c.clusters[newc.UID] = newc
+		c.mu.Unlock()
+	}()
+
+	return c.reconcile(newc)
+}
+
+func (c *Controller) processDelete(ctx context.Context, v interface{}) error {
+	o := v.(*stanv1alpha1.NatsStreamingCluster)
+	log.Infof("Deleted '%s/%s' cluster (uid=%s)", o.Namespace, o.Name, o.UID)
+
+	return nil
+}
+
+func (c *Controller) reconcile(o *stanv1alpha1.NatsStreamingCluster) error {
+	pods, err := c.findRunningPods(o.Name, o.Namespace)
+	if err != nil {
+		return err
+	}
+	n := len(pods) - int(o.Spec.Size)
+	if n == 0 {
+		log.Debugf("Reconciled '%s/%s' cluster (size=%d/%d)", o.Namespace, o.Name, o.Spec.Size, o.Spec.Size)
+		return nil
+	} else if n > 0 {
+		log.Infof("Too many pods for '%s/%s' cluster (size=%d/%d), removing %d pods...", o.Namespace, o.Name, len(pods), o.Spec.Size, n)
+		return c.shrinkCluster(pods, n)
+	} else if n < 0 {
+		log.Infof("Missing pods for '%s/%s' cluster (size=%d/%d), creating %d pods...", o.Namespace, o.Name, len(pods), o.Spec.Size, n*-1)
+
+		// If this is the first time we have perceived the pod,
+		// then try to create the first node to bootstrap the
+		// cluster.
+		c.mu.Lock()
+		if _, ok := c.clusters[o.UID]; !ok {
+			c.mu.Unlock()
+			return c.createBootstrapPod(o)
+		}
+		c.mu.Unlock()
+
+		// If no other nodes are available, then create one with the bootstrap
+		// flag so that it can become the leader.
+		if n == 0 {
+			return c.createBootstrapPod(o)
+		}
+		return c.createMissingPods(o, n*-1)
+	}
+
+	return nil
+}
+
+func (c *Controller) createBootstrapPod(o *stanv1alpha1.NatsStreamingCluster) error {
+	pod := newStanPod(o)
+	pod.Name = fmt.Sprintf("%s-1", o.Name)
+	container := stanContainer(o)
+	container.Command = stanContainerBootstrapCmd(o, pod)
+	pod.Spec.Containers = []k8scorev1.Container{container}
+
+	log.Infof("Creating bootstrap pod '%s/%s'", o.Namespace, pod.Name)
+	_, err := c.kc.CoreV1().Pods(o.Namespace).Create(pod)
+	if err != nil && !k8serrors.IsAlreadyExists(err) {
+		log.Errorf("Failed to create bootstrap Pod: %v", err)
+		return err
+	}
+
+	return nil
+}
+
+func stanContainer(o *stanv1alpha1.NatsStreamingCluster) k8scorev1.Container {
+	container := k8scorev1.Container{
+		Name: "stan",
+	}
+	if o.Spec.Image != "" {
+		container.Image = o.Spec.Image
+	} else {
+		container.Image = DefaultNATSStreamingImage
+	}
+
+	return container
+}
+
+func stanContainerCmd(o *stanv1alpha1.NatsStreamingCluster, pod *k8scorev1.Pod) []string {
+	return []string{
+		"/nats-streaming-server",
+		"-store", "file", "-dir", "store",
+		"-clustered", "-cluster_id", o.Name,
+		fmt.Sprintf("--cluster_node_id=%q", pod.Name),
```

Comment:
Are the node names changing? If so, that may become an issue because as of now, the configuration can't be changed (that is, a peer that goes away will still be attempted to be contacted by RAFT).
It is usually best to define node id names and list the peers, but I understand that it may not be doable here.

Reply:
I did notice that, before the cluster ids were being auto generated but now changed so that the `cluster_id` names in a cluster are always the same, for example for a cluster of size the `cluster_id`s will always be stan-node-1, stan-node-2, stan-node-3...

Reply:
Sounds good.